Add login name to users

GMTA opened the pull request:
This adds a column to the users table which initially contains the UID. The value contained in this column determines the 'username' part of the credentials used for OC. By changing this value, you effectively change the username with which a user can login. This only works for the database backend.

Notes:
* Login name renaming is not supported in either the backends or UI yet, but this is easily added by adding a new backend capability
* The OC upgrade system does not allow for new unique NOT NULL columns to be added to a table. To cope with this, a nullable column is added and populated afterwards by a new repair step. This repair step is not executed on install, but as a new install does not yet have any users, there are no rows that need to be populated.
* If an app always calls ::checkPassword() with the UID of the user instead of his username, this becomes a backward incompatible change. However, the parameter is called $loginname instead of $uid, so you could say the app was already breaking the API contract.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ba073377-fbbf-4d48-a71e-6cae5e86d51e) was created.

GMTA commented:
@owncloud-bot This contribution is MIT licensed.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ba073377-fbbf-4d48-a71e-6cae5e86d51e) completed: **10 new issues, 12 updated code elements**

PVince81 commented:
@blizzz @schiesbn @DeepDiver1975 @icewind1991 please have a look

DeepDiver1975 commented:
I'd consider this change rather critical because it will break any user backend we have out there - refs https://github.com/owncloud/core/pull/14620/files#diff-906e50f731aef7818204adbf47445b24R63

Furthermore it only works around the database mess we have in the area of users and groups.

I honestly prefer to fix the mess and break existing user backends ONLY in this case.

GMTA commented:
You are right about the backend breaking - how about putting the method in OC_User_Backend and defaulting to `return $uid;` ? That way, backends that do not define `->getLoginName()` will keep their previous behavior.

It is true that it works around the DB mess in OC. [I've had to submit PRs to other projects](https://github.com/fmbiete/Z-Push-contrib/commit/803f6303c202efa0fd3e298fd7a0ec1dedb56fc8) in order to get a somewhat professional "feeling" using OC in combination with other solutions. However, I do feel that this is a more backwards compatible way to get the rename functionality in quicker. I'm actually a bit scared to think about refactoring the user backend and UIDs that are used _everywhere_ in OC and the apps...

DeepDiver1975 commented:
> You are right about the backend breaking - how about putting the method in OC_User_Backend and defaulting to return $uid; ? That way, backends that do not define ->getLoginName() will keep their previous behavior.

That would be more acceptable - yes. If you don't mind feel free to walk that path.
But I can still not guarantee that we are at the end :speak_no_evil: 

This is a mess and we need to be careful as this is an integral part of oc. THX a lot!

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/187495f6-7a86-40c5-b964-34dcbc300dd0) was created.

GMTA commented:
Rebased onto master, resolved conflict in repair.php, and removed the `getLoginName` from OC_User_Interface. I already added a default implementation of `getLoginName` to OC_User_Backend so the change was simple.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/187495f6-7a86-40c5-b964-34dcbc300dd0) completed: **12 new issues, 10 updated code elements**

DeepDiver1975 commented:
@GMTA thanks a lot - looks better - but I'm still no convinced. Please allow be to find the time to review this the following days. Thanks a lot for your work! 
I hope this will not discourage you from future contributions! :speak_no_evil:

GMTA commented:
@DeepDiver1975 Thanks and don't worry, I'm dedicated to improving OC's auth/carddav/caldav components in order to get it to a state where I can use it in a professional production environment. As far as I'm concerned, this PR isn't essential anymore as I've [added username mapping support](https://github.com/fmbiete/Z-Push-contrib/commit/803f6303c202efa0fd3e298fd7a0ec1dedb56fc8) to the client side of my architecture.

nickvergessen commented on lib/repair/fixloginname.php:
```diff
@@ -0,0 +1,37 @@
+	public function run() {
+		$qb = $this->connection->createQueryBuilder();
+		$qb->update('`*PREFIX*users`')
+			->set('`loginname`', '`UID`')
```
lowercase column name (`uid`)

nickvergessen commented on lib/private/user.php:
```diff
@@ -522,16 +522,16 @@ public static function canUserChangeDisplayName($uid) {
-		$manager = self::getManager();
-		$username = $manager->checkPassword($uid, $password);
+	public static function checkPassword($loginname, $password) {
+		$manager = \OC::$server->getUserManager();
```
revert this line please?

nickvergessen commented on lib/private/user/database.php:
```diff
@@ -50,8 +50,8 @@ class OC_User_Database extends OC_User_Backend implements \OCP\IUserBackend {
-			$query = OC_DB::prepare('INSERT INTO `*PREFIX*users` ( `uid`, `password` ) VALUES( ?, ? )');
-			$result = $query->execute(array($uid, \OC::$server->getHasher()->hash($password)));
+			$query = OC_DB::prepare('INSERT INTO `*PREFIX*users` ( `uid`, `loginname`, `password` ) VALUES( ?, ?, ? )');
+			$result = $query->execute(array($uid, $uid, \OC::$server->getHasher()->hash($password)));
```
we need to catch the exception when the login name is already taken

GMTA replied:
You're absolutely right about this, and although this can only happen when someone manually edits the loginname for a user in the database, this is something we need to take care of within this PR.

However, this would be a bit late in the process of creating a new user. Duplicate `uid`s are not caught here, so I think `loginname` should also be checked earlier in the process? Do you have a suggestion?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3002e5a0-c46c-4764-9462-f181d5c41165) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f56dd6e2-d28d-4927-afa6-59b799266a1a) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3002e5a0-c46c-4764-9462-f181d5c41165) completed: **9 new issues, 10 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f56dd6e2-d28d-4927-afa6-59b799266a1a) completed: **9 new issues, 10 updated code elements**

Xenopathic commented:
Why do we need the `getLoginName()` method? There is the `$session->get('loginname')` call that gets the same thing, and doesn't need any extra code (and ultimately won't break existing user backends). I like the ability to change the login name in the DB, but I don't like the breakage of compatibility.

GMTA commented:
@Xenopathic What do you think would break existing user backends exactly?

We do not _need_ the `getLoginName()` method at all, I could just hack in `$session->get('loginname')` to get the job done. But even then, why is it necessary to keep the loginname in the session?

As for LDAP, the login name is just the `uid`, just like it is now.

Xenopathic commented:
@GMTA The login name belongs in the session, IMHO, given that there could be multiple login names (only one of which was chosen to login to the current session). And as said before: the LDAP backend has multiple login names, but with `getLoginName()` you would only ever be able to return one. cc @blizzz

GMTA commented:
@Xenopathic Yes I understand, but _why_ do we need the login name in the session? What would happen if we were to remove it? Wouldn't it be enough to have the `uid`?

Xenopathic commented:
There are some circumstances where getting the real login name is useful. I have used it to implement SMB_OC in files_external, which needs the user's username (*not* uid) to attempt to mount an SMB share. In an LDAP environment, this literally means the username that was used to log into oC with, which this PR won't really help with.

GMTA commented:
@Xenopathic That is, of course, because the `loginname`  your referring to is semantically completely different from the `loginname` this PR is suggesting. Would removing all instances of `getLoginName` and replacing them with `$session->get('loginname')` where necessary be agreeable with you?

icewind1991 commented:
Please keep the number of code paths accessing the session directly to a minimum since it makes it harder to modify things in the future

GMTA commented:
@icewind1991 @Xenopathic Then the only option I see (apart from thrashing this entire PR) is to add a capability indicating whether the user backend supports login name retrieval, and only in that case use it for logins. An additional password validation check should be added to accommodate changing your password for user backends that do not support retrieving the login name, as the current password checking mechanism requires the login name.

Xenopathic commented:
My question is, what is the purpose of `getLoginName()`? It won't be used in any internal oC code (that uses the UID always), maybe only when talking to external services, but then you get back to the point of having multiple login names.

My opinion is that this PR should implement the login name DB field only, and not augment the user interface at all.

GMTA commented:
@Xenopathic If a user changes its password, it needs to enter its current password followed by the new password. Password validation in OC is performed on the login name, not the uid, so we need a way to retrieve the login name.

So that's why an alternative would be to introduce another method to validate a user's password based on its `uid`.

blizzz commented:
@GMTA for that use case the login name from the session is sufficient, as it's that one the user used to login with.  The drawback i see is that it might be outdated when the login is changed and the user does not need to login again. As @Xenopathic points out a user may have multiple login names on a backend (e.g. uid and mail on LDAP), i.e. your solution should be flexible to deal with this.

blizzz commented on apps/user_ldap/user_ldap.php:
```diff
@@ -301,6 +301,15 @@ public function getHome($uid) {
+	 * @return string The user's login name
+	 */
+	public function getLoginName($uid) {
+		return $uid;
```
The admin can specify many attributes to be used for login in the login filter and the default scenario is that none of the match the username. To extract attributes for the loginname you would need to parse the the login filter and fetch the values from LDAP. Which means that more than one can be returned.

Xenopathic commented on settings/changepassword/controller.php:
```diff
@@ -8,16 +8,16 @@ public static function changePersonalPassword($args) {
 		$oldPassword = isset($_POST['oldpassword']) ? $_POST['oldpassword'] : '';
 
-		if (!\OC_User::checkPassword($username, $oldPassword)) {
+		if (!\OC_User::checkPassword($user->getLoginName(), $oldPassword)) {
```
The whole reason for `getLoginName()` seems to be this one line. What about changing the logic here instead, so that it passes the session loginname to `checkPassword()` instead of the result of `getUser()`? Then there is no need for a new method on the user backend interface.

GMTA commented:
Guys, I'm abandoning this PR.

Today, I installed a minor OC update (8.0.0 -> 8.0.2) which effectively killed the contacts and calendar apps with a _PHP fatal error_, rendering me without CardDAV / CalDAV. If OC uses anything that remotely resembles semver, then a minor update should have never broken compatibility this badly.

Looking at this discussion and the events of today, I don't think this is anywhere near the right time to introduce a new hackish feature like being able to change the login credentials which - in my humble opinion - should have been designed better to begin with.

I'll try to contribute some of my efforts towards OC quality and getting those apps to work with OC 8.
